### PR TITLE
feat(expo): improve loading expo configs, support app.config.ts

### PR DESCRIPTION
**Pull request by evelant**
### Description

Parse expo configs with `@expo/config`. This adds support for `app.config.ts` and makes projects configured with expo conform to the schema for app.config.ts/js. 

Fixes potential crash in `ios_config.sh` when attempting to parse `.js` file as `.json`. Minor fixes to `app-json.gradle`.

Note: This works on the assumption that `app.config.js/ts` means it's an expo project and safe to load `@expo/config`. I'm not aware of any use of these files other than expo.

### Release Summary

feat(expo): improve loading expo configs, support app.config.ts

### Checklist

- I read the [Contributor Guide](https://github.com/invertase/react-native-google-mobile-ads/blob/main/CONTRIBUTING.md)
  and followed the process outlined there for submitting PRs.
  - [ X ] Yes
- My change supports the following platforms;
  - [ X ] `Android`
  - [ X ] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `__tests__e2e__`
  - [ ] `jest` tests added or updated in `__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ X ] Yes - minor, config moved from root to `"expo.extra"` for expo projects to conform with expo app.config.ts schema.
  - [ ] No

### Test Plan

There aren't any tests for expo configs yet. Tested manually on my production project. `app.config.ts` is resolved and configuration loaded correctly for iOS and Android.

---

Think `react-native-google-mobile-ads` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`Invertase`](https://twitter.com/invertaseio) on Twitter
🔥

### CODE REVIEW

**Comment by evelant** [comment]
I signed the CLA, not sure why the bot doesn't see it

**Comment by VictorioMolina** [comment]
Thank you :D Awesome you included `.ts` support too. You saved my day.

**Comment by evelant** [comment]
Just occurred to me there's a potential for confusion here if someone uses `pnpm` since `@expo/config` won't be hoisted to root node_modules by default so it won't get resolved. Not sure there's much that can be done about that, maybe just a note in the docs that they need to explicitly install `@expo/config`.

**Review by VictorioMolina — APPROVED** [review]
This fixes https://github.com/invertase/react-native-google-mobile-ads/pull/517

**Comment by mikehardy** [comment]
@evelant perhaps a mismatch between commit email address and registered github email addresses?

> evelant seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).


That's all I can think of since `evelant` *is* a github user, but the commits are actually keyed by email IIRC

File: docs/index.mdx [hunk 1]
```diff
@@ -84,13 +84,13 @@ Before you are able to display ads to your users, you must have a [Google AdMob
 "Apps" menu item, create or choose an existing Android/iOS app. Each app platform exposes a unique account ID which needs to
 be added to the project.
 
-> Attempting to build your app without a valid App ID in `app.json` or `app.config.js` will cause the app to crash on start or fail to build.
+> Attempting to build your app without a valid App ID in `app.json` or `app.config.js/ts` will cause the app to crash on start or fail to build.
```
Comment (mikehardy) [inline docs/index.mdx:87]: here we are in backticks so I think we can use regex style to make it clear this is not a directory with a `ts` entry

```suggestion
> Attempting to build your app without a valid App ID in `app.json` or `app.config[.js|.ts]` will cause the app to crash on start or fail to build.
```

File: docs/index.mdx [hunk 2]
```diff
@@ -84,13 +84,13 @@ Before you are able to display ads to your users, you must have a [Google AdMob
...
 
 Under the "App settings" menu item, you can find the "App ID":
 
 ![Google AdMob App ID](https://prismic-io.s3.amazonaws.com/invertase%2F52dd6900-108c-47a6-accb-699fde963b99_new+project+%2813%29.jpg)
 
-Within the root of your React Native project, open the `app.json` or (if using Expo) `app.config.js` file and add the
+Within the root of your React Native project, open the `app.json` or (if using Expo) `app.config.js/ts` file and add the
```
Comment (mikehardy) [inline docs/index.mdx:93]:
```suggestion
Within the root of your React Native project, open the `app.json` or (if using Expo) `app.config[.js|.ts]` file and add the
```

File: docs/index.mdx [hunk 3]
```diff
@@ -127,26 +115,31 @@ It should look like this:
...
   },
-  "react-native-google-mobile-ads": {
-    "android_app_id": "ca-app-pub-xxxxxxxx~xxxxxxxx",
-    "ios_app_id": "ca-app-pub-xxxxxxxx~xxxxxxxx"
-  }
 }
 ```
 
-#### app.config.js
+#### app.config.js/ts
```
Comment (mikehardy) [inline docs/index.mdx:128]:
```suggestion
#### app.config.js/app.config.ts
```

**Comment by evelant** [comment]
Sorry I haven't had a chance to make the requested changes yet. I'll try to get to it soon.

**Comment by evelant** [comment]
Should be good to go now @mikehardy. I improved the ambiguous docs and added a descriptive error message if @expo/config fails to resolve for some reason.

File: ios_config.sh [hunk 4]
```diff
@@ -70,45 +72,47 @@ echo "info: 1) Locating ${_JSON_FILE_NAME} file:"
...
 
-  _CURRENT_LOOKUPS=$((_CURRENT_LOOKUPS+1))
+  _CURRENT_LOOKUPS=$((_CURRENT_LOOKUPS + 1))
 done
 
 if [[ ${_SEARCH_RESULT} ]]; then
   if [[ ${_IS_CONFIG_JS} == "true" ]]; then
-    _JSON_OUTPUT_RAW=$("${NODE_BINARY}" -e "console.log(JSON.stringify(require('${_SEARCH_RESULT}')));")
+    _APP_CONFIG_DIR_NAME=$(dirname $_SEARCH_RESULT)
+    _JSON_OUTPUT_RAW=$("${NODE_BINARY}" -e "try { const expoConfig = require('@expo/config'); console.log(JSON.stringify(expoConfig.getConfig('${_APP_CONFIG_DIR_NAME}')?.exp?.extra));  } catch (e) { if (e instanceof Error && e.code === 'MODULE_NOT_FOUND') { console.error('You have an expo app.config[.js|.ts] file but do not have @expo/config installed. @expo/config is required for react-native-google-mobile-ads to read the app.config[.js|.ts] file and should be installed by default with expo. Please install @expo/config and try again.') } else { throw e }}")
```
Comment (micaeldias) [inline ios_config.sh:105]:
```suggestion
    _JSON_OUTPUT_RAW=$("${NODE_BINARY}" -e "try { const expoConfig = require('@expo/config'); console.log(JSON.stringify(expoConfig.getConfig('${_APP_CONFIG_DIR_NAME}')?.expo?.extra));  } catch (e) { if (e instanceof Error && e.code === 'MODULE_NOT_FOUND') { console.error('You have an expo app.config[.js|.ts] file but do not have @expo/config installed. @expo/config is required for react-native-google-mobile-ads to read the app.config[.js|.ts] file and should be installed by default with expo. Please install @expo/config and try again.') } else { throw e }}")
```

typo?

Comment (evelant) [reply ios_config.sh:105]: Not a typo. The @expo/config package for whatever reason exposes the config under the `exp` key.

**Comment by sairam4123** [comment]
Hey! So I'm using this Pull Request for my own app, and it seems like I'm having this error, 
<details>
<summary>Logs</summary>

```
[eval]:11
Sat, 30 Mar 2024 18:11:22 GMT

          console.error("You have an expo app.config[.js|.ts] file but do not have "@expo/config" installed. "@expo/config" is required for react-native-google-mobile-ads to read the app.config[.js|.ts] file and should be installed by default with expo. Please install "@expo/config" and try again.")
Sat, 30 Mar 2024 18:11:22 GMT

                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Sat, 30 Mar 2024 18:11:22 GMT

SyntaxError: missing ) after argument list
Sat, 30 Mar 2024 18:11:22 GMT

    at new Script (node:vm:100:7)
Sat, 30 Mar 2024 18:11:22 GMT

    at createScript (node:vm:265:10)
Sat, 30 Mar 2024 18:11:22 GMT

    at Object.runInThisContext (node:vm:313:10)
Sat, 30 Mar 2024 18:11:22 GMT

    at node:internal/process/execution:79:19
Sat, 30 Mar 2024 18:11:22 GMT

    at [eval]-wrapper:6:22
Sat, 30 Mar 2024 18:11:22 GMT

    at evalScript (node:internal/process/execution:78:60)
Sat, 30 Mar 2024 18:11:22 GMT

    at node:internal/main/eval_string:28:3
Sat, 30 Mar 2024 18:11:22 GMT

Node.js v18.12.1
Sat, 30 Mar 2024 18:11:22 GMT

FAILURE: Build completed with 2 failures.
Sat, 30 Mar 2024 18:11:22 GMT

1: Task failed with an exception.
Sat, 30 Mar 2024 18:11:22 GMT

-----------
Sat, 30 Mar 2024 18:11:22 GMT

* Where:
Sat, 30 Mar 2024 18:11:22 GMT

Build file '/home/expo/workingdir/build/node_modules/react-native-google-mobile-ads/android/build.gradle' line: 82
Sat, 30 Mar 2024 18:11:22 GMT

* What went wrong:
Sat, 30 Mar 2024 18:11:22 GMT

A problem occurred evaluating project ':react-native-google-mobile-ads'.
Sat, 30 Mar 2024 18:11:22 GMT

> Cannot get property 'googleMobileAdsJson' on extra properties extension as it does not exist
Sat, 30 Mar 2024 18:11:22 GMT

* Try:
Sat, 30 Mar 2024 18:11:22 GMT

> Run with --stacktrace option to get the stack trace.
Sat, 30 Mar 2024 18:11:22 GMT

> Run with --info or --debug option to get more log output.
Sat, 30 Mar 2024 18:11:22 GMT

> Run with --scan to get full insights.
Sat, 30 Mar 2024 18:11:22 GMT

> Get more help at https://help.gradle.org.
Sat, 30 Mar 2024 18:11:22 GMT

==============================================================================
Sat, 30 Mar 2024 18:11:22 GMT

2: Task failed with an exception.
Sat, 30 Mar 2024 18:11:22 GMT

-----------
Sat, 30 Mar 2024 18:11:22 GMT

* What went wrong:
Sat, 30 Mar 2024 18:11:22 GMT

A problem occurred configuring project ':react-native-google-mobile-ads'.
Sat, 30 Mar 2024 18:11:22 GMT

> compileSdkVersion is not specified. Please add it to build.gradle
Sat, 30 Mar 2024 18:11:22 GMT

* Try:
Sat, 30 Mar 2024 18:11:22 GMT

> Run with --stacktrace option to get the stack trace.
Sat, 30 Mar 2024 18:11:22 GMT

> Run with --info or --debug option to get more log output.
Sat, 30 Mar 2024 18:11:22 GMT

> Run with --scan to get full insights.
Sat, 30 Mar 2024 18:11:22 GMT

> Get more help at https://help.gradle.org.
Sat, 30 Mar 2024 18:11:22 GMT

==============================================================================
```
</details>


is there anything I can do to fix this problem??

**Comment by evelant** [comment]
@sairam4123 are you using pnpm? If so the lack of hoisting may make it so that implicitly installed `@expo/config` cannot be resolved. Try installing `@expo/config` explicitly in your `package.json`.

**Comment by sairam4123** [comment]
Hello, I'm using yarn "1.22.x", and I'm trying to build the Android APK using EAS Build, Sent from Android deviceOn 31 Mar 2024 4:37 pm, Andrew M ***@***.***> wrote:
@sairam4123 are you using pnpm? If so the lack of hoisting may make it so that implicitly installed @expo/config cannot be resolved. Try installing @expo/config explicitly in your package.json.

**Comment by sairam4123** [comment]
I am also getting this error.

```
Unable to resolve "./version" from "..\..\node_modules\react-native-google-mobile-ads\src\index.ts"
```

Could you please help me get this working? This has completely stopped the development of the app.

**Comment by sairam4123** [comment]
Hello @evelant, I'm sorry to let you know but I am currently stuck with this issue which caused the development to grind to an complete stop. I have been facing pressure from my clients to get the project done in time but this issue is currently a major blocker, could I possibly expect a bug-fix soon?

Thank you so much for your time and effort!

**Comment by evelant** [comment]
@sairam4123 I'm not sure where that `Unable to resolve "./version"` is coming from, I don't think it's related to this PR. It sounds like maybe you have a configuration problem with your EAS build or package.json such that modules aren't getting resolved correctly. Are you using a monorepo structure?

**Comment by sairam4123** [comment]
> @sairam4123 I'm not sure where that `Unable to resolve "./version"` is coming from, I don't think it's related to this PR. It sounds like maybe you have a configuration problem with your EAS build or package.json such that modules aren't getting resolved correctly. Are you using a monorepo structure?

Yes, I'm using mono repo structure, it seems that version itself is missing from this repo, do we have any options to fix this?
Do you have any idea about the failure in the eas build? Waiting for eas to complete a new build.

**Comment by evelant** [comment]
I'm not really sure. It seems like you've got a module resolution issue due to monorepo hoisting most likely. I'm not a regular maintainer of this package so I don't know the details why it's trying to load that `version` module. I just wanted to contribute back the support for expo configs that I added for my own project.

**Comment by sairam4123** [comment]
I think I have found the issue, it seems to be an issue with the console.error() because if you see the message more clearly, it says that, it needs ) after the argument list. It could be a possibility that the quotes seem to complete the strings and @expo/config seems to be a some sort of a variable to the compiler 


I would love to hear your thoughts on this.

**Comment by sairam4123** [comment]
I have been waiting for an update for two months, @evelant could you please merge my pr fixing the above-mentioned issue? Let me know when you do!

Also, please rebase this branch if possible!

Thank you for your help! 💖

**Pull request closed without merging**